### PR TITLE
[NPU] Upgrade to v0.16.0

### PR DESCRIPTION
**Pull request by gcanlin**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

- It's time to make `OmniNPUModelRunner` inherit from `OmniGPUModelRunner` to reduce the duplicate code.
- This PR also add the installation from main branch on NPU.
- Because vLLM-Ascend havn't release v0.16.0rc1. But we can develop the new feature based on the commit `e2175d9c7e62b437391dfee996b1375674ba7c18`, which has supported vLLM 0.16.0.
- When vLLM-Ascend release v0.16.0rc1, we only need to add some minor changes so it's safe to merge it PR first before v0.16.0rc1.

## Test Plan

- [x] Qwen3-Omni
- [x] Qwen-Image

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please providing the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please pasting the results comparison before and after, or e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by gcanlin** [comment]
cc @hsliuustc0106 @Gaohan123 @tzhouam PTAL. Thanks!

File: vllm_omni/platforms/npu/worker/npu_model_runner.py [hunk 1]
```diff
@@ -1,49 +1,38 @@
...
-
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self._omni_per_req_additional_information: dict[str, dict] | None = None
-        self._omni_num_scheduled_tokens_np: np.ndarray | None = None
-        self._omni_last_model_output: object | None = None
-
+class OmniNPUModelRunner(OmniGPUModelRunner, NPUModelRunner):
     def load_model(self, *args, **kwargs) -> None:
         super().load_model(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (gcanlin) [reply vllm_omni/platforms/npu/worker/npu_model_runner.py]: Good catch! Replace it by `NPUModelRunner.load_model(...)` now

**Comment by hsliuustc0106** [comment]
please wait for vllm upstream 0.16.0 ready

**Comment by gcanlin** [comment]
@hsliuustc0106 GPU CI has passed. Is it time to merge?

File: docs/getting_started/installation/npu/npu.inc.md [hunk 2]
```diff
@@ -50,11 +50,38 @@ For other installation methods (pip installation, building from source, custom D
 
 We are keeping [issue #997](https://github.com/vllm-project/vllm-omni/issues/997) up to date with the aligned versions of vLLM, vLLM-Ascend, and vLLM-Omni, and also outlining the Q1 roadmap there.
 
-# --8<-- [end:installation]
+# --8<-- [end:installation-release]
```
Comment (lishunyang12) [inline docs/getting_started/installation/npu/npu.inc.md:53]: This should be a start marker, not end. Without a matching start:installation-main, the snippet include in npu.md will render as empty.

File: vllm_omni/platforms/npu/worker/npu_model_runner.py [hunk 3]
```diff
@@ -1,49 +1,38 @@
...
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self._omni_per_req_additional_information: dict[str, dict] | None = None
-        self._omni_num_scheduled_tokens_np: np.ndarray | None = None
-        self._omni_last_model_output: object | None = None
-
+class OmniNPUModelRunner(OmniGPUModelRunner, NPUModelRunner):
     def load_model(self, *args, **kwargs) -> None:
-        super().load_model(*args, **kwargs)
+        NPUModelRunner.load_model(self, *args, **kwargs)
```
Comment (lishunyang12) [inline vllm_omni/platforms/npu/worker/npu_model_runner.py:31]: Now that load_model explicitly calls NPUModelRunner.load_model(self, ...), the MRO concern is handled. Worth double-checking: are there other methods where the MRO picks OmniGPUModelRunner's version but you actually need NPUModelRunner's? e.g. _preprocess, _prepare_inputs -- did you verify those resolve correctly at runtime?

File: vllm_omni/platforms/npu/worker/npu_model_runner.py [hunk 4]
```diff
@@ -627,327 +317,74 @@ def dummy_drafter_compute_logits(hidden_states):
...
-            self._merge_additional_information_update(new_req.req_id, payload_info)
-
-        if hasattr(scheduler_output.scheduled_cached_reqs, "additional_information"):
-            cached_infos = getattr(scheduler_output.scheduled_cached_reqs, "additional_information", {})
-            if isinstance(cached_infos, dict):
-                for req_id, req_infos in cached_infos.items():
-                    self._merge_additional_information_update(req_id, req_infos)
-
-    def _preprocess(
+    def _model_forward(
```
Comment (lishunyang12) [inline vllm_omni/platforms/npu/worker/npu_model_runner.py:320]: The new _model_forward takes num_tokens_padded as a positional arg, but the old one did not. Is OmniGPUModelRunner._model_forward also updated to accept num_tokens_padded, or will the GPU path break when called with this extra arg?

File: vllm_omni/platforms/npu/worker/npu_ar_model_runner.py [hunk 5]
```diff
@@ -47,13 +47,14 @@ class ExecuteModelState(NamedTuple):
     scheduler_output: SchedulerOutput
     logits: torch.Tensor
     spec_decode_metadata: SpecDecodeMetadata | None
+    spec_decode_common_attn_metadata: AscendCommonAttentionMetadata | None
```
Comment (lishunyang12) [inline vllm_omni/platforms/npu/worker/npu_ar_model_runner.py:50]: ExecuteModelState namedtuple gained two fields and lost one. Since NamedTuple is positional, mismatches will silently shift values. Would switching to a dataclass be worth it here?

File: vllm_omni/platforms/npu/worker/npu_generation_model_runner.py [hunk 6]
```diff
@@ -106,79 +240,67 @@ def execute_model(
...
+                scheduler_output,
+                None,
+                spec_decode_metadata,
+                spec_decode_common_attn_metadata,
+                None,
+                None,
+                None,
+                attn_metadata,
+                positions,
+                ec_connector_output,
```
Comment (lishunyang12) [inline vllm_omni/platforms/npu/worker/npu_generation_model_runner.py:297]: Passing None for hidden_states and sample_hidden_states into ExecuteModelState -- intentional for the generation runner? sample_tokens unpacks these and could try to index into them. Does the generation path always short-circuit before that?

File: vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py [hunk 7]
```diff
@@ -270,27 +269,6 @@ def forward(
...
-                    _added_batch_dim = True
-
-                # Handle None input_ids
-                if input_ids is None:
-                    input_ids = torch.zeros(
-                        inputs_embeds.shape[1],
-                        dtype=torch.long,
-                        device=thinker_dev,
-                    ).unsqueeze(0)
-                    _added_batch_dim = True
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py:273]: Nice, removing the NPU batch dimension munging. Does the NPU path now handle arbitrary-dim tensors correctly upstream, or is there a risk of shape mismatches with this change?

File: docker/Dockerfile.npu [hunk 8]
```diff
@@ -7,8 +7,9 @@ WORKDIR ${APP_DIR}
 
 COPY . .
 
-# Install vllm-omni with dev dependencies
-RUN pip install --no-cache-dir -e . --no-build-isolation
+RUN sed -i -E 's/^([[:space:]]*)"fa3-fwd==0\.0\.1",/\1# "fa3-fwd==0.0.1",/' pyproject.toml \
```
Comment (lishunyang12) [inline docker/Dockerfile.npu:10]: The sed commands to patch pyproject.toml at build time are fragile -- if indentation or quoting in pyproject.toml changes, these silently fail. Not blocking, just wondering if an npu-specific requirements approach would be more robust.

File: vllm_omni/platforms/npu/stage_configs/qwen3_omni_moe.yaml [hunk 9]
```diff
@@ -7,7 +7,7 @@
 stage_args:
   - stage_id: 0
     runtime:
-      devices: "0,1,2,3"
+      devices: "0,1"
```
Comment (lishunyang12) [inline vllm_omni/platforms/npu/stage_configs/qwen3_omni_moe.yaml:10]: Shrinking from 4 TP to 2 TP and from 6 to 4 NPUs -- is this just updating the default example to match a more common setup, or does the model actually require fewer devices now?

File: vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py (hunk 7 above)
Comment (gcanlin) [reply vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py:273]: We don't need it anymore.

File: docker/Dockerfile.npu (hunk 8 above)
Comment (gcanlin) [reply docker/Dockerfile.npu:10]: It's a workaround for 0.14.0. But we have fixed it in the main branch now.

File: vllm_omni/platforms/npu/stage_configs/qwen3_omni_moe.yaml (hunk 9 above)
Comment (gcanlin) [reply vllm_omni/platforms/npu/stage_configs/qwen3_omni_moe.yaml:10]: Actually we don't need 4 TP for thinker.

File: vllm_omni/platforms/npu/worker/npu_model_runner.py (hunk 3 above)
Comment (gcanlin) [reply vllm_omni/platforms/npu/worker/npu_model_runner.py:31]: YES. They work as our expectation.

File: vllm_omni/platforms/npu/worker/npu_model_runner.py (hunk 4 above)
Comment (gcanlin) [reply vllm_omni/platforms/npu/worker/npu_model_runner.py:320]: Actually it's introduced by vLLM-Ascend instead of omni side. So we don't need to add it into GPU side.

Comment (lishunyang12) [reply vllm_omni/platforms/npu/worker/npu_model_runner.py:320]: i see

File: vllm_omni/platforms/npu/worker/npu_generation_model_runner.py (hunk 6 above)
Comment (gcanlin) [reply vllm_omni/platforms/npu/worker/npu_generation_model_runner.py:297]: It's going to follow GPU behavior. We would never use these fields for generation models actually.

File: vllm_omni/platforms/npu/worker/npu_ar_model_runner.py (hunk 5 above)
Comment (gcanlin) [reply vllm_omni/platforms/npu/worker/npu_ar_model_runner.py:50]: These fields were also following vLLM-Ascend. We don't plan to do it here even if it's possibly necessary.

File: docs/getting_started/installation/npu/npu.inc.md (hunk 2 above)
Comment (gcanlin) [reply docs/getting_started/installation/npu/npu.inc.md:53]: Done. Thanks

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

**Comment by gcanlin** [comment]
@hsliuustc0106 Could we get a first merge? It will really help us develop new features on NPU easily. Thanks!
